rfc/egress-with-ucan.md

```diff
@@ -0,0 +1,159 @@
+
+#### Update
+
+This DID Method does not support updating the DID Document.
+
+#### Deactivate
+
+This DID Method does not support deactivating the DID Document.
+
+## Making content retrievable with a token
+
+A UCAN delegation is used to allow token access to content. Typically, this is performed by an Agent within a Storacha Client, who has the authority to `/space/content/retrieve` from the Space by way of logging into an Account—thus, the delegation chain flows Space → Account → Agent → Token.
+
+```json5
+// UCAN 1.0
+{
+  "iss": "did:key:zAgent",
+  "aud": "did:bearer:abcde12345",
+  "sub": "did:key:zSpace",
+  "cmd": "/space/content/retrieve",
+  "pol": [
+    ["==", ".cid", "bafy...7pcu"],
+  ],
+  "nonce": …,
+  "exp": …
+}
+```
+
+```json5
+// UCAN 0.9
+{
+  "iss": "did:key:zAgent",
+  "aud": "did:bearer:abcde12345",
+  "att": [
+    {
+      "with": "did:key:zSpace",
+      "can": "space/content/retrieve",
+      "nb": {
+        "cid": "bafy...7pcu"
+      },
+    }
+  ],
+  "nnc": …,
+  "exp": …,
+  "prf": […]
+}
+```
+
+The delegation must be available to the Executor at invocation-time. Since the Invoker will be using a token and not speaking UCAN, they will not be able to deliver the proof, so the Executor must have access to it in a store. The Client should therefore invoke `access/delegate` (UCAN 1.0 equivalent TBD) to store the delegation with Storacha.
+
+## The Gateway
+
+To retrieve a piece of content over HTTP, the Retriever will connect to a Gateway. At first, there will be exactly one Gateway: the existing Storacha Gateway (also known as the web3.storage Gateway). Later, as Storage Nodes become more sophisticated, they may decide to run their own Gateways, offering their own direct egress to customers.
+
+The Gateway will offer an HTTP endpoint. Currently, the Storacha Gateway's endpoint takes the form of `https://<cid>.ipfs.w3s.link/`. The Gateway will accept a token as part of the URL. To serve the request, the Gateway will perform the following steps:
+
+1. Look up the Location Commitment for the given CID. If not found, respond with 404 Not Found.
+2. If there is no token given, serve the request through the standard rate-limiter. (This behavior is likely to change in the future.)
+3. If there is a token given (eg, `abcde12345`), build its corresponding DID (eg, `did:bearer:abcde12345`).
+4. Look up all delegations with the token DID as audience.
+5. Attempt to prove the ability to invoke `/space/content/retrieve` on the Space listed in Location Commitment, with the given CID. If more than one Location Commitment is found, attempt each in turn: a CID may be stored in multiple Spaces, and the token may be able to retrieve the content through one and not another.
+6. If no such proof chain is available, respond with 401 Unauthorized. [or 404 Not Found?]
+7. If a proof chain is found, execute the invocation on the Executor.
+8. Using the information in the receipt, fetch the content and proxy it as the response.
```

Comment:
question: How will the customer ID be identified to log the egress event in the Accounting Service? At which step in this process will the customer ID be retrieved and associated with the request for accurate billing?

Reply:
The billing is tied to the Space, so once we have the Location Commitment (which specifies the Space), we can bill the right customer. That probably deserves a callout in here.

Reply:
Got it. Thanks. Yes, it would be great to have that stated in the RFC as well.

Reply:
I think we're missing a step here! we need to look up the "provider" registered with the space, which is a looking in a dynamo table that Freeway does not have access to - I think this needs to be handled either in w3up or w3infra